### PR TITLE
Add support for compiler plugins

### PR DESCRIPTION
**Pull request by lwronski**

### CODE REVIEW

File: modules/build/src/main/scala/scala/build/options/ClassPathOptions.scala [hunk 1]
```diff
@@ -8,7 +8,8 @@ final case class ClassPathOptions(
   extraCompileOnlyJars: Seq[os.Path] = Nil,
   extraSourceJars: Seq[os.Path] = Nil,
   fetchSources: Option[Boolean] = None,
-  extraDependencies: Seq[AnyDependency] = Nil
+  extraDependencies: Seq[AnyDependency] = Nil,
+  compilerPlugins: Seq[AnyDependency] = Nil
```
Comment (alexarchambault) [inline modules/build/src/main/scala/scala/build/options/ClassPathOptions.scala]: Let's rather put `compilerPlugins` in `ScalaOptions` (along with `scalaVersion`, `scalacOptions`, …).

File: modules/cli/src/main/scala/scala/cli/commands/SharedDependencyOptions.scala [hunk 2]
```diff
@@ -15,7 +15,10 @@ final case class SharedDependencyOptions(
   @HelpMessage("")
   @Name("repo")
   @Name("r")
-    repository: List[String] = Nil
+    repository: List[String] = Nil,
+  @Group("Dependency")
```
Comment (alexarchambault) [inline modules/cli/src/main/scala/scala/cli/commands/SharedDependencyOptions.scala]: Group `Scala` would be a better fit I believe.

File: modules/integration/src/test/scala/scala/cli/integration/CompileTests212.scala [hunk 3]
```diff
@@ -3,5 +3,59 @@ package scala.cli.integration
...
+        * The above copyright notice and this permission notice shall be included in all copies or
+        * substantial portions of the Software.
+        *
+        * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+        * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+        * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+        * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+        * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+        * DEALINGS IN THE SOFTWARE.
+        *
```
Comment (alexarchambault) [inline modules/integration/src/test/scala/scala/cli/integration/CompileTests212.scala]: I'm not sure we need to add this… The few lines below are not a "substantial portions of the Software".

File: modules/integration/src/test/scala/scala/cli/integration/CompileTests212.scala [hunk 4]
```diff
@@ -3,5 +3,59 @@ package scala.cli.integration
...
+        * Originally copied from
+        * (https://github.com/typelevel/kind-projector/blob/main/src/test/scala/polylambda.scala)
```
Comment (alexarchambault) [inline modules/integration/src/test/scala/scala/cli/integration/CompileTests212.scala]: Let's put `https://github.com/typelevel/kind-projector/blob/00bf25cef1b7d01d61a3555cccb6cf38fe30e117/src/test/scala/polylambda.scala` instead (got it by pressing Y on the link above, it's less likely to break in the future)

File: modules/integration/src/test/scala/scala/cli/integration/CompileTests212.scala (hunk 3 above)
Comment (lwronski) [reply modules/integration/src/test/scala/scala/cli/integration/CompileTests212.scala]: It's only one fragment from test, it is potential use case of kind-projector library and not "substantial portions of the Software". 

So I deleted the entry of license and added link to the test in kind-project repo.

File: website/docs/compile.md [hunk 5]
```diff
@@ -70,6 +70,13 @@ scala-cli compile Hello.scala -O -deprecation -O -Xlint:infer-any
 
 `-O` accepts both options with the prefixes above and those without such a prefix.
 
+## Scala compiler plugins
+To add compiler plugins dependencis you'd like to use --compiler-plugin:
```
Comment (alexarchambault) [inline website/docs/compile.md]:
```suggestion
Use `--compiler-plugin` to add compiler plugin dependencies:
```

File: modules/cli/src/main/scala/scala/cli/commands/SharedDependencyOptions.scala [hunk 6]
```diff
@@ -15,7 +15,10 @@ final case class SharedDependencyOptions(
   @HelpMessage("")
   @Name("repo")
   @Name("r")
-    repository: List[String] = Nil
+    repository: List[String] = Nil,
+  @Group("Scala")
+  @HelpMessage("Add compiler plugins dependencies")
```
Comment (alexarchambault) [inline modules/cli/src/main/scala/scala/cli/commands/SharedDependencyOptions.scala:22]: Maybe we could add `"plugin"` and `"P"` aliases (via `@Name` annotations), which are more convenient to type than `--compiler-plugin`.

File: website/docs/compile.md [hunk 7]
```diff
@@ -70,6 +70,13 @@ scala-cli compile Hello.scala -O -deprecation -O -Xlint:infer-any
...
+
+```bash
+scala-cli compile Plugin.scala --compiler-plugin org.typelevel:::kind-projector:0.13.2 --scala 2.12.14
```
Comment (alexarchambault) [inline website/docs/compile.md]: NIT (using `Plugin.scala` is confusing IMO, as it's not a plugin itself)
```suggestion
scala-cli compile Main.scala --compiler-plugin org.typelevel:::kind-projector:0.13.2 --scala 2.12.14
```

**Review by alexarchambault — COMMENTED** [review]
(Only added minor comments.)

File: website/docs/compile.md [hunk 8]
```diff
@@ -70,6 +70,13 @@ scala-cli compile Hello.scala -O -deprecation -O -Xlint:infer-any
 
 `-O` accepts both options with the prefixes above and those without such a prefix.
 
+## Scala compiler plugins
+Use `--compiler-plugin` to add compiler plugin dependencies:
```
Comment (alexarchambault) [inline website/docs/compile.md:74]: Seems `--compiler-plugin` was removed. (We should keep it IMO, along with `-P` and `--plugin`.)

**Review by alexarchambault — APPROVED** [review]

**Pull request merged**